src/core/providers/gdal/qgsgdalprovider.cpp

```diff
@@ -715,175 +718,296 @@ bool QgsGdalProvider::readBlock( int bandNo, int xBlock, int yBlock, void *data
+    {
+      if ( GDALRasterIO( gdalBand, GF_Read, 0, srcTop, 1, srcHeight,
+                         static_cast<char *>( data ) + tgtTopOri * bufferWidthPix * dataSize,
+                         1, tgtBottomOri - tgtTopOri + 1, type,
+                         dataSize, dataSize * bufferWidthPix ) != CE_None )
+      {
+        return false;
+      }
+      tgtLeft ++;
+    }
+    if ( reqExtent.yMaximum() - tgtTop * reqYRes > mExtent.yMaximum() )
+    {
+      if ( GDALRasterIO( gdalBand, GF_Read, srcLeft, 0, srcWidth, 1,
+                         static_cast<char *>( data ) + tgtLeftOri * dataSize,
+                         tgtRightOri - tgtLeftOri + 1, 1, type,
+                         dataSize, dataSize * bufferWidthPix ) != CE_None )
+      {
+        return false;
+      }
+      tgtTop ++;
+    }
+    if ( reqExtent.xMinimum() + ( tgtRight + 1 ) * reqXRes > mExtent.xMaximum() )
+    {
+      if ( GDALRasterIO( gdalBand, GF_Read, xSize() - 1, srcTop, 1, srcHeight,
+                         static_cast<char *>( data ) + ( tgtTopOri * bufferWidthPix + tgtRightOri ) * dataSize,
+                         1, tgtBottomOri - tgtTopOri + 1, type,
+                         dataSize, dataSize * bufferWidthPix ) != CE_None )
+      {
+        return false;
+      }
+      tgtRight --;
+    }
+    if ( reqExtent.yMaximum() - ( tgtBottom + 1 ) * reqYRes < mExtent.yMinimum() )
+    {
+      if ( GDALRasterIO( gdalBand, GF_Read, srcLeft, ySize() - 1, srcWidth, 1,
+                         static_cast<char *>( data ) + ( tgtBottomOri * bufferWidthPix + tgtLeftOri ) * dataSize,
+                         tgtRightOri - tgtLeftOri + 1, 1, type,
+                         dataSize, dataSize * bufferWidthPix ) != CE_None )
+      {
+        return false;
+      }
+      tgtBottom --;
+    }
 
-  QgsDebugMsgLevel( QStringLiteral( "width = %1 height = %2 srcWidth = %3 srcHeight = %4" ).arg( width ).arg( height ).arg( srcWidth ).arg( srcHeight ), 5 );
+    int tgtWidth = tgtRight - tgtLeft + 1;
+    int tgtHeight = tgtBottom - tgtTop + 1;
+    if ( tgtWidth > 0 && tgtHeight > 0 )
+    {
+      QgsDebugMsgLevel( QStringLiteral( "using GDAL resampling path" ), 5 );
+      GDALRasterIOExtraArg sExtraArg;
+      INIT_RASTERIO_EXTRA_ARG( sExtraArg );
 
+      const double resamplingFactor = std::max( reqXRes / srcXRes, reqYRes / std::fabs( srcYRes ) );
+      ResamplingMethod method;
+      if ( resamplingFactor < 1 )
+      {
+        method = mZoomedInResamplingMethod;
+      }
+      else if ( resamplingFactor < 1.1 )
+      {
+        // very close to nominal resolution ==> use either zoomed out resampler or zoomed in resampler
+        if ( mZoomedOutResamplingMethod != ResamplingMethod::Nearest )
+          method = mZoomedOutResamplingMethod;
+        else
+          method = mZoomedInResamplingMethod;
+      }
+      else
+      {
+        method = mZoomedOutResamplingMethod;
+      }
+      if ( method == ResamplingMethod::Bilinear )
```

Comment:
Zoomed out bilinear (i.e. the "average" combo box item) doesn't work on my machine:
![Peek 2020-06-12 12-00](https://user-images.githubusercontent.com/1728657/84466788-6fe34200-aca4-11ea-8cf1-6ced0ba755db.gif)

Reply:
> Zoomed out bilinear (i.e. the "average" combo box item) doesn't work on my machine:

It depends on if your image has overviews or not, and the value of the oversampling factor. With an oversampling factor of 2 and if you zoom out by more than a factor of 2 compared to the nominal resolution of the raster, no resampling will occur

Reply:
That'd seem to be a regression from non provider resampling then.

This geotif I used didn't have an overview, I'd expected the zoomed out bilinear resampling with a factor of 2 to work as it used to.

Reply:
@rouault , this is the geotiff I used to conduct some testing: https://storage.googleapis.com/earthenginepartners-hansen/GFC-2019-v1.7/Hansen_GFC-2019-v1.7_treecover2000_20N_100E.tif

Reply:
> With an oversampling factor of 2 and if you zoom out by more than a factor of 2 compared to the nominal resolution of the raster, no resampling will occur

The QGIS logic for its zoomed out resampling is: an oversampling factor of 2 means it'll request a block twice the size and average out the values back into the requested block size. I gather that's not want GDAL does?

Reply:
> The QGIS logic for its zoomed out resampling is: an oversampling factor of 2 means it'll request a block twice the size and average out the values back into the requested block size. I gather that's not want GDAL does?

Yes I saw that QGIS resample filter has this logic, but GDAL doesn't do that: if you zoom out by a factor of 10 and you don't have overviews, then GDAL will need approximatively to fetch 10x10 source pixels to compute an output pixel, which will kill performance. There's no way to have accurate resampling (like GDAL does) and performance in that situation

Reply:
Hmm, that's a significant visual regression / cost.